bokeh/util/api_crawler.py

```diff
@@ -0,0 +1,235 @@
+import ast, os, copy
+
+__all__ = ["api_crawler", "differ"]
+
+
+class APICrawler(object):
+    """
+    Api Crawler: Given a directory, it will use AST to crawl through each of the
+    File and produce AST nodes, which can then be parsed into a representation
+    of the structure of that file. The execution...
+    """
+
+    # Exclude these folders from search.
+    exclude = ("tests", "static", "sampledata", "mplexplorer")
+
+
+    # Point the crawler at this directory.
+    def __init__(self, directory):
+        self.directory = directory
+
+    def is_public(self, name):
+        # Determines whether a given function or class is public or private. Names
+        # with underscores are not allowed unless the name is "__init__"
+        return not name.startswith("_", 0, 1) or name == "__init__"
+
+    def is_function(self, ast_node):
+        # Checks whether a node is a top level function. Checks whether it is an AST
+        # function instance. Also uses column offset, i.e. whether the function has
+        # any tabs or spaces before it.
+        return isinstance(ast_node, ast.FunctionDef) and ast_node.col_offset == 0
+
+    def is_class(self, ast_node):
+        # Checks whether a node is an instance of an ast Class
+        return isinstance(ast_node, ast.ClassDef)
+
+    def get_classes(self, source):
+        # For a given source, walk through the file looking for classes, and assign
+        # the associated methods. Pass them as a dictionary containing each class
+        # and its methods.
+        parsed = ast.parse(source)
+        classes = [node for node in ast.walk(parsed) if self.is_class(node) and self.is_public(node.name)]
+        class_defs = {}
+        for x in classes:
+            class_defs[x.name] = {}
+            methods = []
+            for y in x.body:
+                if isinstance(y, ast.FunctionDef) and self.is_public(y.name):
+                    methods.append(y.name)
+            class_defs[x.name]["methods"] = methods
+        return class_defs
+
+    def get_functions(self, source):
+        # For a given source, look for functions and return those functions as a list.
+        parsed = ast.parse(source)
+        functions = [node for node in ast.walk(parsed) if self.is_function(node) and self.is_public(node.name)]
+        functions = [x.name for x in functions]
+        return functions
+
+    def get_filenames(self, directory):
+        # Go through the file tree and grab the filnames for each file found.
+        # Assign these to an array and return that array.
+        files = []
+        tree = os.walk(directory, topdown=True, followlinks=False)
+        for dirpath, dirnames, filenames in tree:
+            for folder in self.exclude:
+                if folder in dirpath:
```

Comment:
I'm not sure of the how robust (essentially) doing `str.isin` is going to be, in terms of avoiding filename collisions. I could imagine we add a directory named 'static_handler' or similar. We may need to make `self.exclude` contain a real list of directories to ignore instead of trying to match stuff.